File: Java/red-black tree/llama3_3_70b.java
package Java.red-black tree;
import java.util.Random;

class Node {
    int data;
    Node left;
    Node right;
    Node parent;
    boolean isRed;

    public Node(int data) {
        this.data = data;
        this.left = null;
        this.right = null;
        this.parent = null;
        this.isRed = true; // New nodes are always red
    }
}
public class llama3_3_70b {
 
    Node root;
    Node nullNode;

    public llama3_3_70b() {
        nullNode = new Node(0);
        nullNode.isRed = false;
        nullNode.left = null;
        nullNode.right = null;
        root = nullNode;
    }

    // Pre-order helper function
    private void preOrderHelper(Node node) {
        if (node != nullNode) {
            // Perform operation on node
            // For example, you can add the node's data to a list
            preOrderHelper(node.left);
            preOrderHelper(node.right);
        }
    }

    // In-order helper function
    private void inOrderHelper(Node node) {
        if (node != nullNode) {
            inOrderHelper(node.left);
            // Perform operation on node
            // For example, you can add the node's data to a list
            inOrderHelper(node.right);
        }
    }

    // Post-order helper function
    private void postOrderHelper(Node node) {
        if (node != nullNode) {
            postOrderHelper(node.left);
            postOrderHelper(node.right);
            // Perform operation on node
            // For example, you can add the node's data to a list
        }
    }

    // Left rotation function
    private void leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        if (y.left != nullNode) {
            y.left.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.left) {
            x.parent.left = y;
        } else {
            x.parent.right = y;
        }
        y.left = x;
        x.parent = y;
    }

    // Right rotation function
    private void rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        if (y.right != nullNode) {
            y.right.parent = x;
        }
        y.parent = x.parent;
        if (x.parent == null) {
            root = y;
        } else if (x == x.parent.right) {
            x.parent.right = y;
        } else {
            x.parent.left = y;
        }
        y.right = x;
        x.parent = y;
    }

    // Insert function
    public void insert(int data) {
        Node node = new Node(data);
        node.parent = null;
        node.left = nullNode;
        node.right = nullNode;

        Node y = null;
        Node x = root;

        while (x != nullNode) {
            y = x;
            if (node.data < x.data) {
                x = x.left;
            } else {
                x = x.right;
            }
        }

        node.parent = y;
        if (y == null) {
            root = node;
        } else if (node.data < y.data) {
            y.left = node;
        } else {
            y.right = node;
        }

        if (node.parent == null) {
            node.isRed = false;
            return;
        }

        if (node.parent.parent == null) {
            return;
        }

        fixInsert(node);
    }

    // Fix insert function
    private void fixInsert(Node k) {
        Node u;
        while (k.parent.isRed) {
            if (k.parent == k.parent.parent.right) {
                u = k.parent.parent.left;
                if (u.isRed) {
                    u.isRed = false;
                    k.parent.isRed = false;
                    k.parent.parent.isRed = true;
                    k = k.parent.parent;
                } else {
                    if (k == k.parent.left) {
                        k = k.parent;
                        rightRotate(k);
                    }
                    k.parent.isRed = false;
                    k.parent.parent.isRed = true;
                    leftRotate(k.parent.parent);
                }
            } else {
                u = k.parent.parent.right;

                if (u.isRed) {
                    u.isRed = false;
                    k.parent.isRed = false;
                    k.parent.parent.isRed = true;
                    k = k.parent.parent;
                } else {
                    if (k == k.parent.right) {
                        k = k.parent;
                        leftRotate(k);
                    }
                    k.parent.isRed = false;
                    k.parent.parent.isRed = true;
                    rightRotate(k.parent.parent);
                }
            }
            if (k == root) {
                break;
            }
        }
        root.isRed = false;
    }

    // Generate 1000 random inputs and insert them into the Red-Black Tree
    public void generateAndInsertRandomInputs() {
        Random random = new Random();
        for (int i = 0; i < 1000; i++) {
            int randomData = random.nextInt(1001);
            insert(randomData);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        RedBlackTree tree = new RedBlackTree();
        tree.generateAndInsertRandomInputs();
    }
}

File: Python/red-black tree/llama3-3-70b.py

import random

class Node:
    def __init__(self, key, value, color, left=None, right=None, parent=None):
        self.key = key
        self.value = value
        self.color = color
        self.left = left
        self.right = right
        self.parent = parent

class RedBlackTree:
    def __init__(self):
        self.NIL = Node(None, None, 'black')
        self.root = self.NIL

    def pre_order_helper(self, node):
        if node != self.NIL:
            return [node.key] + self.pre_order_helper(node.left) + self.pre_order_helper(node.right)
        return []

    def in_order_helper(self, node):
        if node != self.NIL:
            return self.in_order_helper(node.left) + [node.key] + self.in_order_helper(node.right)
        return []

    def post_order_helper(self, node):
        if node != self.NIL:
            return self.post_order_helper(node.left) + self.post_order_helper(node.right) + [node.key]
        return []

    def search_tree(self, k):
        current = self.root
        while current != self.NIL:
            if current.key == k:
                return current
            if k < current.key:
                current = current.left
            else:
                current = current.right
        return None

    def minimum(self, node):
        while node.left != self.NIL:
            node = node.left
        return node

    def maximum(self, node):
        while node.right != self.NIL:
            node = node.right
        return node

    def successor(self, x):
        if x.right != self.NIL:
            return self.minimum(x.right)
        y = x.parent
        while y != self.NIL and x == y.right:
            x = y
            y = y.parent
        return y

    def predecessor(self, x):
        if x.left != self.NIL:
            return self.maximum(x.left)
        y = x.parent
        while y != self.NIL and x == y.left:
            x = y
            y = y.parent
        return y

    def left_rotate(self, x):
        y = x.right
        x.right = y.left
        if y.left != self.NIL:
            y.left.parent = x
        y.parent = x.parent
        if x.parent == None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        y.left = x
        x.parent = y

    def right_rotate(self, x):
        y = x.left
        x.left = y.right
        if y.right != self.NIL:
            y.right.parent = x
        y.parent = x.parent
        if x.parent == None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        y.right = x
        x.parent = y

    def insert_node(self, key, value):
        node = Node(key, value, 'red')
        node.parent = None
        node.left = self.NIL
        node.right = self.NIL

        y = None
        x = self.root

        while x != self.NIL:
            y = x
            if node.key < x.key:
                x = x.left
            else:
                x = x.right

        node.parent = y
        if y == None:
            self.root = node
        elif node.key < y.key:
            y.left = node
        else:
            y.right = node

        if node.parent == None:
            node.color = 'black'
            return

        if node.parent.parent == None:
            return

        self.fix_insert(node)

    def fix_insert(self, k):
        while k.parent.color == 'red':
            if k.parent == k.parent.parent.right:
                u = k.parent.parent.left
                if u.color == 'red':
                    u.color = 'black'
                    k.parent.color = 'black'
                    k.parent.parent.color = 'red'
                    k = k.parent.parent
                else:
                    if k == k.parent.left:
                        k = k.parent
                        self.right_rotate(k)
                    k.parent.color = 'black'
                    k.parent.parent.color = 'red'
                    self.left_rotate(k.parent.parent)
            else:
                u = k.parent.parent.right

                if u.color == 'red':
                    u.color = 'black'
                    k.parent.color = 'black'
                    k.parent.parent.color = 'red'
                    k = k.parent.parent
                else:
                    if k == k.parent.right:
                        k = k.parent
                        self.left_rotate(k)
                    k.parent.color = 'black'
                    k.parent.parent.color = 'red'
                    self.right_rotate(k.parent.parent)
            if k == self.root:
                break
        self.root.color = 'black'

    def delete_node(self, key):
        z = self.search_tree(key)
        if z == None:
            return
        y = z
        y_original_color = y.color
        if z.left == self.NIL:
            x = z.right
            self.rb_transplant(z, z.right)
        elif z.right == self.NIL:
            x = z.left
            self.rb_transplant(z, z.left)
        else:
            y = self.minimum(z.right)
            y_original_color = y.color
            x = y.right
            if y.parent == z:
                x.parent = y
            else:
                self.rb_transplant(y, y.right)
                y.right = z.right
                y.right.parent = y

            self.rb_transplant(z, y)
            y.left = z.left
            y.left.parent = y
            y.color = z.color
        if y_original_color == 'black':
            self.delete_fix(x)

    def rb_transplant(self, u, v):
        if u.parent == None:
            self.root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        v.parent = u.parent

    def delete_fix(self, x):
        while x != self.root and x.color == 'black':
            if x == x.parent.left:
                s = x.parent.right
                if s.color == 'red':
                    s.color = 'black'
                    x.parent.color = 'red'
                    self.left_rotate(x.parent)
                    s = x.parent.right

                if s.left.color == 'black' and s.right.color == 'black':
                    s.color = 'red'
                    x = x.parent
                else:
                    if s.right.color == 'black':
                        s.left.color = 'black'
                        s.color = 'red'
                        self.right_rotate(s)
                        s = x.parent.right

                    s.color = x.parent.color
                    x.parent.color = 'black'
                    s.right.color = 'black'
                    self.left_rotate(x.parent)
                    x = self.root
            else:
                s = x.parent.left
                if s.color == 'red':
                    s.color = 'black'
                    x.parent.color = 'red'
                    self.right_rotate(x.parent)
                    s = x.parent.left

                if s.right.color == 'black' and s.right.color == 'black':
                    s.color = 'red'
                    x = x.parent
                else:
                    if s.left.color == 'black':
                        s.right.color = 'black'
                        s.color = 'red'
                        self.left_rotate(s)
                        s = x.parent.left

                    s.color = x.parent.color
                    x.parent.color = 'black'
                    s.left.color = 'black'
                    self.right_rotate(x.parent)
                    x = self.root
        x.color = 'black'

# Test the implementation
tree = RedBlackTree()
keys = [random.randint(1, 100) for _ in range(1000)]
for key in keys:
    tree.insert_node(key, key)

# Test search
for key in keys:
    assert tree.search_tree(key).key == key

# Test delete
for key in keys:
    tree.delete_node(key)
    assert tree.search_tree(key) is None

# Test insert and delete with random keys
for _ in range(1000):
    key = random.randint(1, 100)
    tree.insert_node(key, key)
    assert tree.search_tree(key).key == key
    tree.delete_node(key)
